From 701886b278bd2825a3b5b8cdba7bd83901beca37 Mon Sep 17 00:00:00 2001
From: Nick Sullivan <nick.sullivan@adhocteam.us>
Date: Thu, 7 Jan 2021 11:28:15 -0500
Subject: [PATCH] [Resources and Support] Add 'browse by audience' section into
 homepage (#15541)

* [resources and support] Add 'browse by audience' into homepage

* little clean up

* include non_bene in audience list

* ad some templating

* add the 'show more' button

* add beta banner

* add into build
---
 .../layouts/basic_landing_page.drupal.liquid  | 52 +++++++++++++++++++
 .../drupal/graphql/GetAllPages.graphql.js     |  2 +
 .../graphql/GetLatestPageById.graphql.js      |  2 +
 .../GetTaxonomies.graphql.js                  | 21 ++++++++
 src/site/stages/build/drupal/page.js          | 31 +++++++++++
 5 files changed, 108 insertions(+)
 create mode 100644 src/site/stages/build/drupal/graphql/taxonomy-fragments/GetTaxonomies.graphql.js

diff --git a/src/site/layouts/basic_landing_page.drupal.liquid b/src/site/layouts/basic_landing_page.drupal.liquid
index a77368c615..9279ac6508 100644
--- a/src/site/layouts/basic_landing_page.drupal.liquid
+++ b/src/site/layouts/basic_landing_page.drupal.liquid
@@ -50,6 +50,58 @@
             {% include "src/site/includes/support_resources_search_bar.drupal.liquid" %}
           </div>
 
+          <div class="usa-alert usa-alert-info" role="alert">
+            <div class="usa-alert-body">
+                <h3 class="usa-alert-heading">
+                    We're currently in beta testing
+                </h3>
+                <div class="processed-content">
+                  <div itemprop="text">
+                    <p>Welcome to resources and support, a new part of <a href="/">VA.gov</a>. We'll be adding more articles and topics in the weeks and months ahead, so please check back often.</p>
+                  </div>
+                </div>
+            </div>
+          </div>
+
+          {% if audienceTags != empty %}
+            <h2>Browse by audience</h2>
+            <ul class="usa-unstyled-list">
+
+              {% assign pageSize = 5 %}
+
+              {% for audienceTag in audienceTags %}
+                {% if forloop.index <= pageSize %}
+                  <li class="vads-u-margin-y--2">
+                    <a class="vads-u-font-weight--bold vads-u-text-decoration--none" href="{{ audienceTag.path.alias }}/">
+                      {{ audienceTag.name }}
+                      <i role="presentation" aria-hidden="true" class="fa fa-chevron-right vads-u-margin-left--1"> </i>
+                    </a>
+                  </li>
+                {% endif %}
+              {% endfor %}
+
+              {% if audienceTags.length > pageSize %}
+                <div class="form-expanding-group additional-info-container vads-u-margin-y--2p5">
+                  <div class="additional-info-title">Show more</div>
+                  {% assign remainingAudienceTags = audienceTags | sliceArrayFromStart: pageSize %}
+                  <span>
+                      <div class="additional-info-content">
+                        <ul class="usa-unstyled-list">
+                          {% for audienceTag in remainingAudienceTags %}
+                            <li class="vads-u-margin-y--2">
+                              <a class="vads-u-font-weight--bold vads-u-text-decoration--none" href="{{ audienceTag.path.alias }}/">
+                                {{ audienceTag.name }}
+                                <i role="presentation" aria-hidden="true" class="fa fa-chevron-right vads-u-margin-left--1"> </i>
+                              </a>
+                            </li>
+                          {% endfor %}
+                        </ul>
+                      </div>
+                  </span>
+              </div>
+            {% endif %}
+          {% endif %}
+
           <!-- Content blocks -->
           {% for block in fieldContentBlock %}
             {% assign bundleComponent = "src/site/paragraphs/" | append: block.entity.entityBundle %}
diff --git a/src/site/stages/build/drupal/graphql/GetAllPages.graphql.js b/src/site/stages/build/drupal/graphql/GetAllPages.graphql.js
index c41041bacf..bbd3213086 100644
--- a/src/site/stages/build/drupal/graphql/GetAllPages.graphql.js
+++ b/src/site/stages/build/drupal/graphql/GetAllPages.graphql.js
@@ -33,6 +33,7 @@ const qaPage = require('./nodeQa.graphql');
 const sidebarQuery = require('./navigation-fragments/sidebar.nav.graphql');
 const stepByStepPage = require('./nodeStepByStep.graphql');
 const storyListingPage = require('./storyListingPage.graphql');
+const taxonomiesQuery = require('./taxonomy-fragments/GetTaxonomies.graphql');
 const supportResourcesDetailPage = require('./nodeSupportResourcesDetailPage.graphql');
 const vaFormPage = require('./vaFormPage.graphql');
 const vamcOperatingStatusAndAlerts = require('./vamcOperatingStatusAndAlerts.graphql');
@@ -154,6 +155,7 @@ const buildQuery = ({ useTomeSync }) => {
         : ''
     }
     ${menuLinksQuery}
+    ${taxonomiesQuery}
   }
 `;
 
diff --git a/src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js b/src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js
index 9fe82e62cd..49c17cff6a 100644
--- a/src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js
+++ b/src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js
@@ -24,6 +24,7 @@ const nodeStepByStep = require('./nodeStepByStep.graphql');
 const nodeSupportResourcesDetailPage = require('./nodeSupportResourcesDetailPage.graphql');
 const pressReleasePage = require('./pressReleasePage.graphql');
 const sidebarQuery = require('./navigation-fragments/sidebar.nav.graphql');
+const taxonomiesQuery = require('./taxonomy-fragments/GetTaxonomies.graphql');
 const vaFormPage = require('./vaFormPage.graphql');
 const vamcOperatingStatusAndAlerts = require('./vamcOperatingStatusAndAlerts.graphql');
 
@@ -102,6 +103,7 @@ module.exports = `
         : ''
     }
     ${menuLinksQuery}
+    ${taxonomiesQuery}
   }
 `;
 
diff --git a/src/site/stages/build/drupal/graphql/taxonomy-fragments/GetTaxonomies.graphql.js b/src/site/stages/build/drupal/graphql/taxonomy-fragments/GetTaxonomies.graphql.js
new file mode 100644
index 0000000000..44ead9baea
--- /dev/null
+++ b/src/site/stages/build/drupal/graphql/taxonomy-fragments/GetTaxonomies.graphql.js
@@ -0,0 +1,21 @@
+module.exports = `
+  allTaxonomies: taxonomyTermQuery(limit: 1000) {
+    entities {
+      entityBundle
+      ... on TaxonomyTermAudienceNonBeneficiaries {
+        fieldAudienceRsHomepage
+        name
+        path {
+          alias
+        }
+      }
+      ... on TaxonomyTermAudienceBeneficiaries {
+        fieldAudienceRsHomepage
+        name
+        path {
+          alias
+        }
+      }
+    }
+  }
+`;
diff --git a/src/site/stages/build/drupal/page.js b/src/site/stages/build/drupal/page.js
index 2f4e87d14c..6557551231 100644
--- a/src/site/stages/build/drupal/page.js
+++ b/src/site/stages/build/drupal/page.js
@@ -248,6 +248,27 @@ function getFacilitySidebar(page, contentData) {
   return { links: [] };
 }
 
+function mergeTaxonomiesIntoResourcesAndSupportHomepage(
+  resourcesAndSupportHomepage,
+  allTaxonomies,
+) {
+  const audienceBundles = new Set([
+    'audience_beneficiaries',
+    'audience_non_beneficiaries',
+  ]);
+
+  const audienceTagsUnsorted = allTaxonomies.entities
+    .filter(taxonomy => audienceBundles.has(taxonomy.entityBundle))
+    .filter(audienceTag => audienceTag.fieldAudienceRsHomepage);
+
+  const audienceTags = _.sortBy(audienceTagsUnsorted, 'name');
+
+  return {
+    ...resourcesAndSupportHomepage,
+    audienceTags,
+  };
+}
+
 function compilePage(page, contentData) {
   const {
     data: {
@@ -265,6 +286,9 @@ function compilePage(page, contentData) {
       alerts: alertsItem = {},
       bannerAlerts: bannerAlertsItem = {},
       outreachSidebarQuery: outreachSidebarNav = {},
+      allTaxonomies = {
+        entities: [],
+      },
     },
   } = contentData;
 
@@ -388,6 +412,13 @@ function compilePage(page, contentData) {
       break;
   }
 
+  if (entityUrl.path === '/resources') {
+    pageCompiled = mergeTaxonomiesIntoResourcesAndSupportHomepage(
+      pageCompiled,
+      allTaxonomies,
+    );
+  }
+
   return pageCompiled;
 }
 
-- 
2.29.2

